src/common/FileSystem.cpp

```diff
@@ -905,6 +910,74 @@ class ZipArchive {
 			ClearErrorCode(err);
 	}
 
+	// OpenFile but with support for symlinks.
+	// Symlinks are a bad feature which you should not use. Therefore, the implementation is as
+	// slow as possible with a full iteration of the archive performed for each symlink.
+	// Returns: Length of the opened file, if successful.
+	offset_t OpenFileWithSymlinkResolution(Str::StringRef name, offset_t offset, std::error_code& err)
+	{
+		int depth = 0;
+#ifdef BUILD_ENGINE
+		int maxDepth = fs_maxSymlinkDepth.Get();
+#else
+		int maxDepth = 0; // TODO(slipher): Stop building unzip code in the gamelogic.
+#endif
+		std::string resolvedName;
+		for (;;) {
+			OpenFile(offset, err);
+			if (err)
+				return 0;
+			unz_file_info64 fileInfo;
+			int result = unzGetCurrentFileInfo64(zipFile, &fileInfo, nullptr, 0, nullptr, 0, nullptr, 0);
+			if (result != UNZ_OK) {
+				SetErrorCodeZlib(err, result);
+				return 0;
+			}
+			if (!IsSymlink(fileInfo)) {
+				// success
+				return fileInfo.uncompressed_size;
+			}
+			if (depth >= maxDepth) {
+				fsLogs.Warn("Failed to open symlinked zip archive file '%s': maximum depth of %d exceeded", name, depth);
+				CloseFile(err);
+				if (err)
+					return 0;
+				// TODO: find better error code?
+				SetErrorCodeFilesystem(err, filesystem_error::no_such_file);
+				return 0;
+			}
+			++depth;
+			char link[MAX_FILENAME_BUF];
+			size_t linkLength = ReadFile(&link, sizeof(link) - 1, err);
+			if (err) {
+				// If there was a read error, are we still supposed to close the file? Oh well.
+				return 0;
+			}
+			CloseFile(err);
+			if (err)
+				return 0;
+			link[linkLength] = '\0';
+			resolvedName = ResolveLinkPath(name, link, err);
+			if (err)
+				return 0;
+			name = resolvedName;
+			if (!Path::IsValid(name, /*allowDir=*/false)) {
+				fsLogs.Warn("Symlink resolved to invalid filename '%s'", name);
+				SetErrorCodeFilesystem(err, filesystem_error::invalid_filename);
+				return 0;
+			}
+			auto maybeOffset = FindOffsetForName(name, err);
+			if (err)
+				return 0;
+			if (!maybeOffset) {
+				fsLogs.Warn("Symlink points to nonexistent file '%s'", name);
+				SetErrorCodeFilesystem(err, filesystem_error::no_such_file);
+				return 0;
+			}
+			offset = *maybeOffset;
```

Comment:
This would please me a lot:

```diff
diff --git a/src/common/FileSystem.cpp b/src/common/FileSystem.cpp
index f88c31ee5..85fd981a6 100644
--- a/src/common/FileSystem.cpp
+++ b/src/common/FileSystem.cpp
@@ -960,20 +960,20 @@ public:
                        resolvedName = ResolveLinkPath(name, link, err);
                        if (err)
                                return 0;
-                       name = resolvedName;
-                       if (!Path::IsValid(name, /*allowDir=*/false)) {
-                               fsLogs.Warn("Symlink resolved to invalid filename '%s'", name);
+                       if (!Path::IsValid(resolvedName, /*allowDir=*/false)) {
+                               fsLogs.Warn("Symlink resolved to invalid filename '%s' → '%s'", name, resolvedName);
                                SetErrorCodeFilesystem(err, filesystem_error::invalid_filename);
                                return 0;
                        }
-                       auto maybeOffset = FindOffsetForName(name, err);
+                       auto maybeOffset = FindOffsetForName(resolvedName, err);
                        if (err)
                                return 0;
                        if (!maybeOffset) {
-                               fsLogs.Warn("Symlink points to nonexistent file '%s'", name);
+                               fsLogs.Warn("Symlink points to nonexistent file '%s' → '%s'", name, resolvedName);
                                SetErrorCodeFilesystem(err, filesystem_error::no_such_file);
                                return 0;
                        }
+                       fsLogs.Debug("Symlink resolved: '%s' → '%s'", name, resolvedName);
                        offset = *maybeOffset;
                }
        }
```